jupyterhub/apihandlers/base.py

```diff
@@ -62,6 +85,43 @@ def check_referer(self):
             return False
         return True
 
+    def get_scope_filter(self, req_scope):
```

Comment:
We can use [lru_cache()](https://docs.python.org/3/library/functools.html#functools.lru_cache) to memoize this since we call it so many times (once each for a few scopes for each user in user_model) with the same few arguments:

```python
@lru_cache()
def get_scope_filter(self, req_scope):
```